### PR TITLE
Introduce a Proof-of-Validation block type and use that in place of BlockData

### PR DESCRIPTION
**Pull request by rphmeier**
Most of the time when we use `BlockData` we should have been using this `PoVBlock` type, which ties `BlockData` together with `ConsolidatedIngress`. This shifts the burden of fetching parachain ingress entirely to collators, who send it along with their collations to the validators. 

There are further optimizations that can be done by omitting the ingress if the validator has signalled that they've seen it already, but they are beyond the scope of this PR.

### CODE REVIEW

File: network/src/validation.rs [hunk 1]
```diff
@@ -746,22 +747,31 @@ impl<P: ProvideRuntimeApi + Send, E, N, T> SessionDataFetcher<P, E, N, T> where
...
 {
-	/// Fetch block data for the given candidate receipt.
-	pub fn fetch_block_data(&self, candidate: &CandidateReceipt) -> BlockDataReceiver {
+	/// Fetch PoV block for the given candidate receipt.
+	pub fn fetch_pov_block(&self, candidate: &CandidateReceipt) -> PoVReceiver {
+		let parachain = candidate.parachain_index.clone();
 		let parent_hash = self.parent_hash;
+
+		let canon_roots = self.api.runtime_api().ingress(&BlockId::hash(parent_hash), parachain)
+			.map_err(|e| format!("Cannot fetch ingress for parachain {:?} at {:?}: {:?}",
```
Comment (bkchr) [inline network/src/validation.rs]: "weird" formatting.

File: primitives/src/parachain.rs [hunk 2]
```diff
@@ -159,26 +159,52 @@ impl Ord for CandidateReceipt {
...
+
+/// Parachain ingress queue message.
+#[derive(PartialEq, Eq, Clone, Decode)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Encode, Debug))]
 pub struct Message(#[cfg_attr(feature = "std", serde(with="bytes"))] pub Vec<u8>);
 
+/// Consolidated ingress roots.
+///
+/// This is an ordered vector of other parachains' egress queue roots,
+/// obtained according tothe routing rules. The same parachain may appear
```
Comment (bkchr) [inline primitives/src/parachain.rs]:
```suggestion
/// obtained according to the routing rules. The same parachain may appear
```

File: validation/src/collation.rs [hunk 3]
```diff
@@ -145,15 +146,33 @@ error_chain! {
 			display("Collated for inactive parachain: {:?}", id),
 		}
 		EgressRootMismatch(id: ParaId, expected: Hash, got: Hash) {
-			description("Got unexpected egress route."),
+			description("Got unexpected egress rooe."),
```
Comment (bkchr) [inline validation/src/collation.rs]:
```suggestion
			description("Got unexpected egress root."),
```

File: network/src/validation.rs [hunk 4]
```diff
@@ -746,22 +747,31 @@ impl<P: ProvideRuntimeApi + Send, E, N, T> SessionDataFetcher<P, E, N, T> where
...
+				parachain, parent_hash, e)
+			);
+
+		// TODO: use canon roots in fetching logic.
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply network/src/validation.rs]: (actually is irrelevant)

File: network/src/router.rs [hunk 5]
```diff
@@ -234,26 +241,21 @@ impl<P: ProvideRuntimeApi + Send, E, N, T> TableRouter for Router<P, E, N, T> wh
 	E: Future<Item=(),Error=()> + Clone + Send + 'static,
 {
 	type Error = io::Error;
-	type FetchCandidate = validation::BlockDataReceiver;
-	type FetchIncoming = validation::IncomingReceiver;
+	type FetchValidationProof = validation::PoVReceiver;
 
-	fn local_candidate(&self, receipt: CandidateReceipt, block_data: BlockData, extrinsic: Extrinsic) {
+	fn local_collation(&self, collation: Collation, extrinsic: Extrinsic) {
```
Comment (bkchr) [inline network/src/router.rs]: What is the extrinsic here?

Comment (rphmeier) [reply network/src/router.rs]: it's parachain-extrinsic data (outgoing message queues & balance uploads), not an extrinsic. Terminology is from the Polkadot paper.

**Pull request merged**